src/Language/Fixpoint/Defunctionalize.hs

```diff
@@ -199,10 +202,10 @@ choices (xs:xss) = [a:as | a <- xs, as <- choices xss]
 -- | Containers defunctionalization --------------------------------------------
 --------------------------------------------------------------------------------
 
-class Defunc a where
-  defunc :: a -> DF a
+class Defunc a s | a -> s where
```

Comment:
Do we need this functional dependency? It seems like it compiles without it.

Reply:
I thought we should enable functional dependency whenever possible since it reduces the number of explicit annotations. Is there any negative effect caused by functional dependency that we should avoid?

Reply:
I guess it depends on the instances, but I don't think `a` will always imply an `s`.